index.js

```diff
@@ -460,3 +439,94 @@ function wrapPromise() {
     }
   }
 }
+
+// Shim activator for functions that have callback last
+function activator(fn) {
+  var fallback = function () {
+    var index = arguments.length - 1;
+    if (typeof arguments[index] === "function") {
+      arguments[index] = wrapCallback(arguments[index]);
+    }
+    return fn.apply(this, arguments);
+  };
+  // Preserve function length for small arg count functions.
+  switch (fn.length) {
+    case 1:
+      return function (cb) {
+        if (arguments.length !== 1) return fallback.apply(this, arguments);
+        if (typeof cb === "function") cb = wrapCallback(cb);
```

Comment:
Cool, thanks for finding that.

On Sat, Mar 12, 2016, 10:11 PM Matthew Loring <notifications@github.com>
wrote:

> In index.js
> <https://github.com/othiym23/async-listener/pull/54#discussion_r55932695>:
>
> > +
> > +// Shim activator for functions that have callback last
> > +function activator(fn) {
> > +  var fallback = function () {
> > +    var index = arguments.length - 1;
> > +    if (typeof arguments[index] === "function") {
> > +      arguments[index] = wrapCallback(arguments[index]);
> > +    }
> > +    return fn.apply(this, arguments);
> > +  };
> > +  // Preserve function length for small arg count functions.
> > +  switch (fn.length) {
> > +    case 1:
> > +      return function (cb) {
> > +        if (arguments.length !== 1) return fallback.apply(this, arguments);
> > +        if (typeof cb === "function") cb = wrapCallback(cb);
>
> It appears that reassigning argument variables is only a performance
> concern in non-strict mode. See:
> https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#3-managing-arguments
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/othiym23/async-listener/pull/54/files#r55932695>.
>